phpBB/phpbb/composer/extension_manager.php

```diff
@@ -0,0 +1,316 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\composer;
+
+use Composer\IO\IOInterface;
+use phpbb\cache\driver\driver_interface;
+use phpbb\composer\exception\managed_with_clean_error_exception;
+use phpbb\composer\exception\managed_with_enable_error_exception;
+use phpbb\composer\exception\runtime_exception;
+use phpbb\config\config;
+use phpbb\extension\manager as ext_manager;
+use phpbb\filesystem\exception\filesystem_exception;
+use phpbb\filesystem\filesystem;
+
+/**
+ * Class to safely manage extensions through composer.
+ */
+class extension_manager extends manager
+{
+	/**
+	 * @var \phpbb\extension\manager
+	 */
+	protected $extension_manager;
+
+	/**
+	 * @var \phpbb\filesystem\filesystem
+	 */
+	protected $filesystem;
+
+	/**
+	 * @var array
+	 */
+	private $enabled_extensions;
+
+	/**
+	 * @var bool Enables extensions when installing them?
+	 */
+	private $enable_on_install = false;
+
+	/**
+	 * @var bool Purges extensions data when removing them?
+	 */
+	private $purge_on_remove = true;
+
+	/**
+	 * @param installer			$installer			Installer object
+	 * @param driver_interface	$cache				Cache object
+	 * @param ext_manager		$extension_manager	phpBB extension manager
+	 * @param filesystem		$filesystem			Filesystem object
+	 * @param string			$package_type		Composer type of managed packages
+	 * @param string			$exception_prefix	Exception prefix to use
+	 * @param string			$root_path			phpBB root path
+	 * @param config			$config				Config object
+	 */
+	public function __construct(installer $installer, driver_interface $cache, ext_manager $extension_manager, filesystem $filesystem, $package_type, $exception_prefix, $root_path, config $config = null)
+	{
+		$this->extension_manager = $extension_manager;
+		$this->filesystem = $filesystem;
+		$this->root_path = $root_path;
+
+		if ($config)
+		{
+			$this->enable_on_install = (bool) $config['exts_composer_enable_on_install'];
+			$this->purge_on_remove   = (bool) $config['exts_composer_purge_on_remove'];
+		}
+
+		parent::__construct($installer, $cache, $package_type, $exception_prefix);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function pre_install(array $packages, IOInterface $io = null)
+	{
+		$installed_manually = array_intersect(array_keys($this->extension_manager->all_available()), array_keys($packages));
+		if (count($installed_manually) !== 0)
```

Comment:
To be honest, we might want to change that to count() anyway in the coding guidelines, especially since sizeof() might be confusing for people coming from other programming languages.